src/format/KeePass2Repair.cpp

```diff
@@ -84,21 +75,20 @@ KeePass2Repair::RepairResult KeePass2Repair::repairDatabase(QIODevice* device, c
     KeePass2XmlReader xmlReader;
     QBuffer buffer(&xmlData);
     buffer.open(QIODevice::ReadOnly);
-    xmlReader.readDatabase(&buffer, db, &randomStream);
+    xmlReader.readDatabase(&buffer, db.data(), &randomStream);
 
     if (xmlReader.hasError()) {
-        delete db;
         return RepairFailed;
     }
     else {
-        m_db = db;
+        m_db.reset(db.take());
         return RepairSuccess;
     }
 }
 
-Database* KeePass2Repair::database() const
+Database* KeePass2Repair::takeDatabase()
 {
-    return m_db;
+    return m_db.take();
```

Comment:
Sorry to nag again, but I didn't see that you need to share ownership of the pointer. In this case, this should be a QSharedPointer instead of a QScopedPointer.
This method should stay const and return the  QSharedPointer instead of a raw pointer.

Reply:
I agree, but I wanted to leave `DatabaseOpenWidget` intact for now.

Reply:
If we change something, we should do it properly from the start.

Reply:
Ok, got it. My goal was to enable `leak_check_at_exit` ASAN check for non-gui tests, without breaking too many interfaces, and then work on other memory management issues.
Considering shared ownership: There's actually no need for `KeePass2Repair` to store the database (pointer, to be exact). What do you think about returning a tuple from `repairDatabase()`, consisting of `RepairResult` and `Database*`?

Reply:
I understand that. I would do the same. But KeePassXC has inherited quite some baggage from KeePassX that needs cleaning up, so in cases like this we should do things properly instead of producing more patchwork.
I had a deeper look into the repair class and agree with you: there is no need to store the database. Returning a `QPair<RepairResult, QSharedPointer<Database>>` sounds like a decent solution to me.

While you are at it could you also please add
```
Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
```
to the license headers? This is missing somehow.

Reply:
Alright, I'll fix it more thoroughly (and will add the copyright notice :slightly_smiling_face:).

Reply:
Thanks a lot!

Reply:
I've investigated that a bit more and I guess my knowledge about KeePassXC is not sufficient for now to clean up the `Database` instances management. Mixing `QSharedPointer<Database>` with raw `Database*` owning pointers is very risky on large scale. On the other hand, reasoning about the `Database` owners and users requires deeper understanding of dependencies, which I lack.

In this particular case, returning `QSharedPointer` from `KeePass2Repair::repairDatabase()` or `KeePass2Repair::database()` forces you to store this object in `DatabaseWidget` to prevent object deletion on zeroing the reference counter. But, replacing `Database* m_db;` with  `QSharedPointer<Database> m_db` causes domino effect on the whole codebase.
``` bash
$ grep "Database\*" -Ir keepassxc/src | wc -l
199
```

On the other hand, using `QScopedPointer<Database>` and transferring ownership explicitly from `KeePass2Repair` to `DatabaseRepairWidget` is not perfect, but still better than raw owning pointer and does not require major changes in `Database` entities management.

Anyway, I won't insist on merging those changes if you disagree.

Reply:
Fair point. Returning a QSharedPointer would indeed require changes everywhere where this pointer is used or require at least explicit transfer of ownership after the repair function, in which case that function may already return a raw pointer.
We will need to refactor raw pointer usage at some point, but then I think we can skip this for now. We plan to redo much of the database handling code in a future release anyway to streamline and encapsulate the code better and not hand around references which should not be of any concern to outside classes.

As for now, I would like to ask you to change the `repairDatabase()` function in a way that it returns a `QPair<RepairResult, Database*>` and remove the `takeDatabase()` method and the `m_db` member variable. That should already be an improvement over the current situation and prevent us from dereferencing `m_db` after ownership has been transferred. When you've done that, please rebase this PR against `release/2.2.2`, so I can merge it into the next patch release.

Reply:
I have done some work already to move to shared pointers for database objects. Its a bear, we should save that for another day.

Reply:
Ok, thanks for assistance and understanding :) I've just pushed fixed `KeePass2Repair`. Just to be clear: should this PR be merged to `release/2.2.2` branch only or to both, `develop` and `release/2.2.2`? I'm not familiar with release/branching policy yet.

Reply:
Only to release/2.2.3 (not 2.2.2). That branch will be merged back into develop then. Just rebase against it and change the target of this PR, I'll take care of the rest.
I hope you also don't mind if I squash your commits into one?

Reply:
I'm a bit stuck in leaky TestGroup.cpp unit tests after rebase. I'll deliver it soon.

Reply:
Alright, the changes are rebased against `release/2.2.3`, squashed and green on CI :smiley:

Reply:
Thanks!